tests/websocket.c

```diff
@@ -8,319 +8,537 @@
-			(int)(total_bytes / (mili1-mili0)));
-		return EXIT_SUCCESS;
-	} else {
-		printf("No message was read.\n");
-		return EXIT_FAILURE;
-	}
+    struct timespec t0, t1;
+
+    int messages_default = 2500;
+    int thread_count_default = 4;
+    short port_default = 7379;
+    char *host_default = "127.0.0.1";
+
+    int msg_target = messages_default;
+    int thread_count = thread_count_default;
+    int i, opt;
+    char *colon;
+    long total = 0, total_bytes = 0;
+    int verbose = 0;
+    int timeout_seconds = -1;
+
+    struct host_info hi = {host_default, port_default};
+
+    struct worker_thread *workers;
+
+    /* getopt */
+    struct option long_options[] = {
+        {"help", no_argument, NULL, '?'},
+        {"host", required_argument, NULL, 'h'},
+        {"port", required_argument, NULL, 'p'},
+        {"clients", required_argument, NULL, 'c'},
+        {"messages", required_argument, NULL, 'n'},
+        {"max-time", required_argument, NULL, 't'},
+        {"verbose", no_argument, NULL, 'v'},
+        {0, 0, 0, 0}
+    };
+    while ((opt = getopt_long(argc, argv, "h:p:c:n:t:vs", long_options, NULL)) != -1) {
+        switch (opt) {
+            case 'h':
+                colon = strchr(optarg, ':');
+                if(!colon) {
+                    size_t sz = strlen(optarg);
+                    hi.host = calloc(1 + sz, 1);
+                    strncpy(hi.host, optarg, sz);
+                } else {
+                    hi.host = calloc(1+colon-optarg, 1);
+                    strncpy(hi.host, optarg, colon-optarg);
+                    hi.port = (short)atol(colon+1);
+                }
+                break;
+
+            case 'p':
+                hi.port = (short)atol(optarg);
+                break;
+
+            case 'c':
+                thread_count = atoi(optarg);
+                break;
+
+            case 'n':
+                msg_target = atoi(optarg);
+                break;
+
+            case 't':
+                timeout_seconds = atoi(optarg);
+                break;
+
+            case 'v':
+                verbose = 1;
+                break;
+
+            default: /* '?' */
+                usage(argv[0], host_default, port_default,
+                        thread_count_default,
+                        messages_default);
+                exit(EXIT_SUCCESS);
+        }
+    }
+
+    /* run threads */
+    workers = calloc(sizeof(struct worker_thread), thread_count);
+
+    clock_gettime(CLOCK_MONOTONIC, &t0);
+    for (i = 0; i < thread_count; ++i) {
+        workers[i].id = i;
+        workers[i].msg_target = msg_target;
+        workers[i].hi = &hi;
+        workers[i].verbose = verbose;
+        workers[i].state = WS_INITIAL;
+        workers[i].debug = verbose ? debug_verbose : debug_noop;
+        workers[i].timeout_seconds = timeout_seconds;
+        if (thread_count == 1) {
+            printf("Single-threaded mode\n");
+            worker_main(&workers[0]);
```

Comment:
I can't tell why this would make a difference vs calling `pthread_create` with a single thread. Debugging… maybe? I'm not even sure this matters.